pkg/ingressCache/pathTrie.go

```diff
@@ -0,0 +1,122 @@
+package ingressCache
+
+import (
+	"fmt"
+	"github.com/dghubble/trie"
+)
+
+const (
+	emptyPathError         = "path is empty"
+	emptyFunctionNameError = "function name is empty"
+	walkPathResultError    = "value is not a []string"
+	functionNotExistsError = "function does not exist in path"
+)
+
+// newPathTree creates a new PathTree instance
+func newPathTree() *pathTree {
+	return &pathTree{*trie.NewPathTrie()}
+}
+
+type pathTree struct {
+	trie.PathTrie
+	// TODO - think about injecting a logger here
+}
+
+func (p *pathTree) set(path string, function string) error {
+	if path == "" {
+		return fmt.Errorf(emptyPathError)
+	}
+
+	if function == "" {
+		return fmt.Errorf(emptyFunctionNameError)
+	}
+
+	// get the exact path value in order to avoid creating a new path if it does not exist
+	pathValue := p.Get(path)
+	if pathValue == nil {
+		p.Put(path, []string{function})
+		return nil
+	}
+
+	pathFunctionNames, ok := pathValue.([]string)
+	if !ok {
+		return fmt.Errorf("%s, got %T", walkPathResultError, pathValue)
+	}
+
+	if existsInSlice(pathFunctionNames, function) {
+		// TODO - add log that function already exists
+		return nil
+	}
+
+	pathValue = append(pathFunctionNames, function)
+	p.Put(path, pathValue)
+
+	return nil
+}
+
+// delete removes a function from a path and also deletes the path if the function is the only one associated with that path
+func (p *pathTree) delete(path string, function string) error {
+	pathValue := p.Get(path)
+
+	pathFunctionNames, ok := pathValue.([]string)
+	if !ok {
+		return fmt.Errorf("%s, got %T", functionNotExistsError, pathValue)
+	}
+
+	// If the function is the only value, delete the path
+	if len(pathFunctionNames) == 1 {
+		if pathFunctionNames[0] == function {
+			p.Delete(path)
+			return nil
+		}
+		// TODO - add log that function does not exist in the path
+		return fmt.Errorf("%s,  function:%s, path:%s", functionNotExistsError, function, path)
```

Comment:
is this really an error?
I would say the "Deletion" succeeded (idempotency - remember)?